pyop2/codegen/rep2loopy.py

```diff
@@ -85,6 +91,210 @@ def petsc_function_lookup(target, identifier):
+
+        return (self.copy(name_in_target=name_in_target,
+                          arg_id_to_dtype={-1: NumpyType(mat_dtype, target=loopy.CTarget()), 0: NumpyType(mat_dtype, target=loopy.CTarget()), 1: NumpyType(numpy.int32, target=loopy.CTarget())}),
+                callables_table)
+
+    def emit_call_insn(self, insn, target, expression_to_code_mapper):
+        assert self.is_ready_for_codegen()
+
+        assert isinstance(insn, loopy.CallInstruction)
+
+        parameters = insn.expression.parameters
+
+        parameters = list(parameters)
+        par_dtypes = [self.arg_id_to_dtype[i] for i, _ in enumerate(parameters)]
+
+        parameters.append(insn.assignees[-1])
+        par_dtypes.append(self.arg_id_to_dtype[0])
+
+        mat_descr = self.arg_id_to_descr[0]
+        arg_c_parameters = [
+            expression_to_code_mapper(
+                par,
+                PREC_NONE,
+                dtype_to_type_context(target, par_dtype),
+                par_dtype
+            ).expr
+            if isinstance(par, SubArrayRef) else
+            expression_to_code_mapper(
+                par,
+                PREC_NONE,
+                dtype_to_type_context(target, par_dtype),
+                par_dtype
+            ).expr
+            for par, par_dtype in zip(parameters, par_dtypes)
+        ]
+        c_parameters = []
+        c_parameters.insert(0, arg_c_parameters[-1])  # t1
+        c_parameters.insert(1, arg_c_parameters[0])  # t0
+        c_parameters.insert(2, numpy.int32(mat_descr.shape[0]))  # n
+        return var(self.name_in_target)(*c_parameters), False
+
+    def generate_preambles(self, target):
+        assert isinstance(target, CTarget)
+        inverse_preamble = """
+            #include <string.h>
+            #include <stdio.h>
+            #include <stdlib.h>
+            #ifndef Inverse_HPP
+            #define Inverse_HPP
+            #define BUF_SIZE 30
+
+            static PetscBLASInt ipiv_buffer[BUF_SIZE];
+            static PetscScalar work_buffer[BUF_SIZE*BUF_SIZE];
+            static void inverse(PetscScalar* restrict Aout, const PetscScalar* restrict A, const PetscBLASInt N)
+            {
+                PetscBLASInt info;
+                PetscBLASInt *ipiv = N <= BUF_SIZE ? ipiv_buffer : malloc(N*sizeof(*ipiv));
+                PetscScalar *Awork = N <= BUF_SIZE ? work_buffer : malloc(N*N*sizeof(*Awork));
+                memcpy(Aout,A,N*N*sizeof(PetscScalar));
+                LAPACKgetrf_(&N,&N,Aout,&N,ipiv,&info);
+                if(info==0)
+                    LAPACKgetri_(&N,Aout,&N,ipiv,Awork,&N,&info);
+                if(info!=0)
+                    fprintf(stderr,\"Getri throws nonzero info.\");
+                if ( N > BUF_SIZE ) {
+                    free(Awork);
+                    free(ipiv);
+                }
+            }
+            #endif
+        """
+        yield("inverse", "#include <petscsys.h>\n#include <petscblaslapack.h>\n"+inverse_preamble)
```

Comment:
style, space before `(`.

Reply:
What? You mean around the +, no?

Reply:
`yield ("inverse", ...)`, not `yield("inverse", ...)`

Reply:
ah